cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/RemoteProgramRunDispatcher.java

```diff
@@ -118,13 +115,18 @@ public ProgramController dispatchProgram(ProgramRunDispatcherInfo programRunDisp
         programId.getType());
     RunRecordDetail runRecordDetail = store.getRun(programRunId);
     TwillController twillController = null;
-    if (remoteTwillRunnerService != null && runRecordDetail != null) {
+    if (runRecordDetail != null) {
       twillController = remoteTwillRunnerService.createTwillControllerFromRunRecord(runRecordDetail);
     }
     ProgramController programController = null;
-    if (runner instanceof DistributedWorkflowProgramRunner && twillController != null) {
-      programController =
-        ((DistributedWorkflowProgramRunner) runner).createProgramController(programRunId, twillController);
+    if (runner instanceof ProgramControllerCreator && twillController != null) {
+      programController = ((ProgramControllerCreator) runner).createProgramController(programRunId, twillController);
+    }
+    if (Objects.isNull(programController)) {
+      String msg = String.format("Unable to dispatch Program %s with runid %s",
+                                 programRunDispatcherInfo.getProgramDescriptor().getProgramId(),
+                                 programRunDispatcherInfo.getRunId());
+      throw Throwables.propagate(new Throwable(msg));
```

Comment:
throw an exception here instead of this.

Reply:
Thanks for catching this. This was a mistake.